src/rebuild/srv.c

```diff
@@ -1457,36 +1463,51 @@ regenerate_task_internal(struct ds_pool *pool, struct pool_target *tgts,
 	return DER_SUCCESS;
 }
 
-/* Regenerate the rebuild tasks when changing the leader. */
 int
-ds_rebuild_regenerate_task(struct ds_pool *pool)
+regenerate_task_of_type(struct ds_pool *pool, pool_comp_state_t match_states,
+			daos_rebuild_opc_t rebuild_op)
 {
 	struct pool_target	*tgts;
 	unsigned int		tgts_cnt;
 	int			rc;
 
-	rebuild_gst.rg_abort = 0;
-
-	/* get all down targets */
-	rc = pool_map_find_down_tgts(pool->sp_map, &tgts, &tgts_cnt);
+	rc = pool_map_find_tgts_by_state(pool->sp_map, match_states,
+					 &tgts, &tgts_cnt);
 	if (rc != 0) {
-		D_ERROR("failed to create failed tgt list rc "DF_RC"\n",
-			DP_RC(rc));
+		D_ERROR("failed to create %s tgt_list rc="DF_RC"\n",
+			RB_OP_STR(rebuild_op), DP_RC(rc));
 		return rc;
 	}
 
-	rc = regenerate_task_internal(pool, tgts, tgts_cnt, RB_OP_FAIL);
+	return regenerate_task_internal(pool, tgts, tgts_cnt, rebuild_op);
+}
+
+
+/* Regenerate the rebuild tasks when changing the leader. */
+int
+ds_rebuild_regenerate_task(struct ds_pool *pool)
+{
+	int rc;
+
+	rebuild_gst.rg_abort = 0;
+
+	rc = regenerate_task_of_type(pool, PO_COMP_ST_DOWN, RB_OP_FAIL);
 	if (rc != 0)
 		return rc;
 
-	/* get all up targets */
-	rc = pool_map_find_up_tgts(pool->sp_map, &tgts, &tgts_cnt);
-	if (rc != 0) {
-		D_ERROR("failed to create up tgt list rc %d\n", rc);
+	rc = regenerate_task_of_type(pool, PO_COMP_ST_DRAIN, RB_OP_DRAIN);
```

Comment:
Maybe just find pull out all targets with DOWN|DRAIN|UP states, then regenerate the tasks by their state.  So it does not need to call regenerate_task_of_type 3 times?

Reply:
I think this logic will be improved later for DAOS-2843 (prioritize rebuild/extend/reint/drain)